activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -4,26 +4,44 @@ module ActiveRecord
   module Associations
     class Preloader
       class Association #:nodoc:
-        attr_reader :preloaded_records
-
         def initialize(klass, owners, reflection, preload_scope)
           @klass         = klass
           @owners        = owners
           @reflection    = reflection
           @preload_scope = preload_scope
           @model         = owners.first && owners.first.class
-          @preloaded_records = []
         end
 
-        def run(preloader)
-          records = load_records do |record|
-            owner = owners_by_key[convert_key(record[association_key_name])]
-            association = owner.association(reflection.name)
-            association.set_inverse_instance(record)
+        def run
+          if !preload_scope || preload_scope.empty_scope?
+            owners.each do |owner|
+              associate_records_to_owner(owner, records_by_owner[owner] || [])
+            end
+          else
+            # Custom preload scope is used and
+            # the association can not be marked as loaded
+            # Loading into a Hash instead
+            records_by_owner
```

Comment:
Is this explict call necessary?

Reply:
No, the only motivation for it was that lazy execution of the database query can change queries order and make it hard to control. However, none of the tests is failing when I comment it out. Do you think it is better to remove it?

Reply:
My preference is keeping accessors as before (`preloaded_records`, and add new `records_by_owner`), and extracting building method. What do you think?

```ruby
        def run
          records_by_owner = load_records # or build_records_by_owner or something else

          if !preload_scope || preload_scope.empty_scope?
            owners.each do |owner|
              associate_records_to_owner(owner, records_by_owner[owner] || [])
            end
          end

          self
        end

        def load_records
          @preloaded_records = ...
          @records_by_owner = ...
        end
```

Reply:
I have a bad experience with this pattern where a lot of ivars are assigned within the same method. It makes it hard to maintain - in case we would need to just access `preloaded_records` but not anything else in the future, and hard to follow: with separated methods implementing the data loading the new person looking at the code can trace the dependencies and execution sequence easily, while in your proposal ivar usage has indirect link to var assignment. It can become even worse: this implementation leads to the same ivar  assigning in multiple places (unless someone supervises it precisely which is very unlikely for rails repo).

In my projects, I broke down multiple ivars assignment in the same method into multiple short methods with cache with clear execution chain and ability to use one by one many times. However, according to current situation, your proposal looks ok. So, I'll leave it up to you. Please just let me know.

Reply:
Okey, I'd leave it as is for now, we could refactor around preloader in any future round.